### PR TITLE
Add Constant Wrappers

### LINKED ISSUE #10: Write specs/tests for SymEngine constants
Check [Beginner Contributor Guide - Ruby Extensions](https://github.com/symengine/symengine/wiki/Beginner-Contributor-Guide---Ruby-Extensions) for the workflow that we follow.

This PR https://github.com/symengine/symengine/pull/781 adds the c wrappers.

This will also involve coming up with an easy-to-use and an intuitive interface for the constants (like pi, E, I) in the gem.

### PR DESCRIPTION
**Pull request by rajithv**
Fixes #10, #11

### CODE REVIEW

File: ext/symengine/ruby_constants.h [hunk 1]
```diff
@@ -0,0 +1,19 @@
+#ifndef RUBY_CONSTANTS_H_
+#define RUBY_CONSTATNS_H_
+
+#include <ruby.h>
+#include <symengine/cwrapper.h>
+
+#include "symengine_macros.h"
+
+void cbasic_free_ptr(void *ptr);
```
Comment (isuruf) [inline ext/symengine/ruby_constants.h]: You can use `cbasic_free_heap` function in `ruby_basic.h`

File: ext/symengine/ruby_constants.h [hunk 2]
```diff
@@ -0,0 +1,19 @@
+#ifndef RUBY_CONSTANTS_H_
+#define RUBY_CONSTATNS_H_
```
Comment (isuruf) [inline ext/symengine/ruby_constants.h]: `RUBY_CONSTATNS_H_` should be changed to `RUBY_CONSTANTS_H_`

File: ext/symengine/symengine.h [hunk 3]
```diff
@@ -11,5 +11,6 @@ VALUE c_basic;
 VALUE c_symbol;
 VALUE c_integer;
 VALUE c_rational;
+VALUE c_constants;
```
Comment (isuruf) [inline ext/symengine/symengine.h]: Can you assign to this variable in `symengine.c`?
Also you have to add it to `Klass_of_basic` method

Comment (rajithv) [inline ext/symengine/symengine.h]: If I make an empty class and assign it to `c_constants` in `symengine.c`, that would be fine?
and set `Klass_of_basic` to the same as well?

Comment (isuruf) [inline ext/symengine/symengine.h]: Yes, right now `E` is an object of type `Basic` which is fine, but you can make a new class `c_constants` and make that the object type of `E`

**Comment by isuruf** [comment]
Due to a recent change in the main repo's travis script, travis-ci exists early.
Can you change this line https://github.com/symengine/symengine.rb/blob/b18621690aa20cb9cc1be0528851516547e3b3cd/.travis.yml#L41 to `- bin/test_travis.sh`

File: ext/symengine/symengine.c [hunk 4]
```diff
@@ -46,4 +47,10 @@ void Init_symengine() {
     c_rational = rb_define_class_under(m_symengine, "Rational", c_basic);
     rb_define_alloc_func(c_rational, cbasic_alloc);
     rb_define_method(c_rational, "initialize", crational_init, 1);
+
+    //Constants
+    c_constants = rb_define_class_under(m_symengine, "Constants", c_basic);
+    rb_define_const(m_symengine, "PI", cconstants_pi());
+    rb_define_const(m_symengine, "E", cconstants_e());
+    rb_define_const(m_symengine, "EulerGamma", cconstants_euler_gamma());
```
Comment (abinashmeher999) [inline ext/symengine/symengine.c]: `EulerGamma` should be in all caps to show that it is a constant.

File: ext/symengine/symengine.c [hunk 5]
```diff
@@ -46,4 +47,10 @@ void Init_symengine() {
     c_rational = rb_define_class_under(m_symengine, "Rational", c_basic);
     rb_define_alloc_func(c_rational, cbasic_alloc);
     rb_define_method(c_rational, "initialize", crational_init, 1);
+
+    //Constants
+    c_constants = rb_define_class_under(m_symengine, "Constants", c_basic);
```
Comment (abinashmeher999) [inline ext/symengine/symengine.c]: This should be singular as in `Constant`. While comparing this sounds intuitive
```ruby
SymEngine::E.instance_of? SymEngine::Constant
```
not `SymEngine::Constants`

**Comment by abinashmeher999** [comment]
Everything else looks good. Maybe after this PR you can add a `cconstant_init`, that lets the user make his/her constants.

File: ext/symengine/symengine.c (hunk 4 above)
Comment (rajithv) [inline ext/symengine/symengine.c]: I was using the notation used at [SymEngine CWrapper](https://github.com/symengine/symengine/blob/master/symengine/cwrapper.cpp#L106). Should I go ahead and change to All Caps or leave it as it is.

Comment (abinashmeher999) [inline ext/symengine/symengine.c]: I would have suggested to stay with the style used in SymEngine but it is a convention in Ruby to name your constants in all caps. See the convention used for constants [here](http://itsignals.cascadia.com.au/?p=7).

File: ext/symengine/CMakeLists.txt [hunk 6]
```diff
@@ -3,6 +3,7 @@ set(RUBY_WRAPPER_SRC
     ruby_symbol.c
     ruby_integer.c
     ruby_rational.c
+	ruby_constants.c
```
Comment (isuruf) [inline ext/symengine/CMakeLists.txt]: Use 4 spaces instead of tabs

File: ext/symengine/ruby_constants.c [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+    VALUE result;
+
+    cresult = basic_new_heap();
+    cwfunc_ptr(cresult);
+    
+    result = Data_Wrap_Struct(c_constant, NULL , cbasic_free_heap, cresult);
+    return result;
+}
+
+VALUE cconstant_pi(){
```
Comment (isuruf) [inline ext/symengine/ruby_constants.c]: Add space before `{`

File: ext/symengine/ruby_constants.c [hunk 8]
```diff
@@ -0,0 +1,24 @@
+#include "ruby_constants.h"
+
+VALUE cconstant_const(void (*cwfunc_ptr)(basic_struct*)){
```
Comment (isuruf) [inline ext/symengine/ruby_constants.c]: Add space before `{`

File: ext/symengine/ruby_constants.c [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+    basic_struct *cresult;
+    VALUE result;
+
+    cresult = basic_new_heap();
+    cwfunc_ptr(cresult);
+    
+    result = Data_Wrap_Struct(c_constant, NULL , cbasic_free_heap, cresult);
```
Comment (isuruf) [inline ext/symengine/ruby_constants.c]: Remove space after `NULL`

File: ext/symengine/symengine.c (hunk 5 above)
Comment (isuruf) [inline ext/symengine/symengine.c]: @rajithv, can you add the statement in the above comment as a test?
Also something like `(x+E-x)instance_of? SymEngine::Constant`

File: spec/constants_spec.rb [hunk 10]
```diff
@@ -0,0 +1,72 @@
+require 'spec_helper'
+
+describe SymEngine do
+  before :each do
+  end
+
+  describe 'Constants' do
```
Comment (abinashmeher999) [inline spec/constants_spec.rb]: This one should be singular.

**Comment by abinashmeher999** [comment]
Could you please change the file names too. `constants` to `constant`.

**Comment by isuruf** [comment]
Looks good to me. @abinashmeher999, any more comments?

**Comment by abinashmeher999** [comment]
Same here. +1 to merge.

**Comment by isuruf** [comment]
@rajithv, thanks for the PR

**Comment by abinashmeher999** [comment]
@rajithv Thanks for your patience.

**Comment by rajithv** [comment]
@isuruf @abinashmeher999 Thank you for the advices as well.

**Comment by certik** [comment]
Thanks for the patch @rajithv.

**Pull request merged**